File: 1138-grumpy-bookstore-owner/grumpy-bookstore-owner.ts
function maxSatisfied(customers: number[], grumpy: number[], minutes: number): number {
     let sum : number = customers.reduce((prev, cur, i) => {
        if(i < minutes) return prev + cur;
        else return grumpy[i] ? prev + 0 : prev + cur;
    }, 0);

    let maxVal : number = sum;

    for(let i = 0; i + minutes < customers.length; i += 1){
        sum = grumpy[i] ? sum - customers[i] : sum;
        sum = grumpy[minutes + i] ? sum + customers[minutes + i] : sum;
        if(maxVal < sum) maxVal = sum;
    };

    return maxVal;
};

File: 1138-grumpy-bookstore-owner/grumpy-bookstore-owner.cpp
class Solution {
public:
    int maxSatisfied(vector<int>& customers, vector<int>& grumpy, int minutes) {
        int n = customers.size();
        int ans = 0;
        int i;
        
        vector<int> p(n, 0);
        vector<vector<int>> dp(n, {2, 0});
        
        int k = customers[0];
        p[0] = grumpy[0] * customers[0]; 
        dp[0][0] = p[0]; 

        for (i = 1; i < n; i++) {
            k += customers[i]; 
            if (grumpy[i] == 1) p[i] = customers[i] + p[i - 1];

            p[i] = max(p[i - 1], p[i]);
            dp[i][0] = p[i];

            if (i >= minutes) {
                dp[i][1] = dp[i - 1][1] + p[i] - p[i - 1]; 
                dp[i][1] = min(dp[i][1], dp[i - minutes][0]); 
            }
        }

        return k - dp[n - 1][1];
    }
};